src/state_awset.erl

```diff
@@ -0,0 +1,468 @@
+        RemovedDotCloud ->
+            ordsets:fold(
+              fun(Dot0, DecompList0) ->
+                      lists:append(
+                        DecompList0,
+                        [{?TYPE,
+                          {state_causal_type:new_data_store({dot_map, dot_set}),
+                           [Dot0]}}])
+              end, DecompList, RemovedDotCloud)
+    end.
+
+%% @private
+add_elem_delta(Elem,
+               Actor,
+               {DataStore0, DotCloud0},
+               {DeltaDataStore0, DeltaDotCloud0}) ->
+    NewDotContext = state_causal_type:get_next_dot_context(Actor, DotCloud0),
+
+    {ok, {dot_set, DotSet}} = state_causal_type:get_sub_data_store(Elem, DataStore0),
+    DeltaDotCloud1 = state_causal_type:insert_dot_context(NewDotContext, DotSet),
+    DeltaDotCloud = state_causal_type:merge_dot_clouds(DeltaDotCloud0,
+                                                       DeltaDotCloud1),
+    DeltaDataStore =
+        state_causal_type:insert_object(Elem,
+                                        {dot_set, DeltaDotCloud1},
+                                        DeltaDataStore0),
+
+    DotCloud = state_causal_type:insert_dot_context(NewDotContext, DotCloud0),
+
+    {ok, {DataStore0, DotCloud}, {DeltaDataStore, DeltaDotCloud}}.
+
+%% @private
+remove_elem_delta(Elem, DataStore, {DeltaDataStore0, DeltaDotCloud0}) ->
+    {ok, SubDataStore} = state_causal_type:get_sub_data_store(Elem, DataStore),
+    case state_causal_type:is_bottom_data_store(SubDataStore) of
+        false ->
+            DeltaDotCloud = state_causal_type:merge_dot_clouds(
+                              DeltaDotCloud0,
+                              state_causal_type:get_dot_cloud(SubDataStore)),
+            {ok, DataStore, {DeltaDataStore0, DeltaDotCloud}};
+        true ->
+            {error, {precondition, {not_present, Elem}}}
+    end.
+
+%% @private
+remove_elems_delta([], DataStore, Delta) ->
+    {ok, DataStore, Delta};
+remove_elems_delta([Elem|Rest], DataStore, {DeltaDataStore0, DeltaDotCloud0}) ->
+    case remove_elem_delta(Elem, DataStore, {DeltaDataStore0, DeltaDotCloud0}) of
+        {ok, DataStore, Delta} -> remove_elems_delta(Rest, DataStore, Delta);
+        Error         -> Error
+    end.
+
+%% ===================================================================
+%% EUnit tests
+%% ===================================================================
+-ifdef(TEST).
+
+new_test() ->
+    ?assertEqual({?TYPE, {{{dot_map, dot_set}, orddict:new()}, ordsets:new()}},
+                 new()).
+
+query_test() ->
+    Set0 = new(),
+    Set1 = {?TYPE, {{{dot_map, dot_set}, [{<<"a">>, {dot_set, [{a, 2}]}}]},
+                    [{a, 1}, {a, 2}]}},
+    ?assertEqual(sets:new(), query(Set0)),
+    ?assertEqual(sets:from_list([<<"a">>]), query(Set1)).
+
+delta_add_test() ->
+    Actor = 1,
+    Set0 = new(),
+    {ok, {?TYPE, {delta, Delta1}}} = delta_mutate({add, <<"a">>}, Actor, Set0),
+    Set1 = merge({?TYPE, Delta1}, Set0),
```

Comment:
Ok, I am assuming that the delta_mutate will just produce a delta that then needs to be joined into the source state to mimic a standard mutation. Looks like it is like that here.